File: Python/day10/AOC.py
from collections import defaultdict

def find_paths(input):
    starts = [(r, c) for r in range(len(input)) for c in range(len(input[0])) if input[r][c] == "0"]
    all_paths = []
    for start in starts:
       all_paths.extend(dfs(input, start, 0, []))
    return all_paths

def dfs(input, pos, num, path):
    i, j = pos
    if input[i][j] != str(num): return []
    path = path + [pos]
    if num == 9:
        return [path]
    paths = []
    for new_pos in [(i-1, j), (i+1, j), (i, j-1), (i, j+1)]:
        if 0 <= new_pos[0] < len(input) and 0 <= new_pos[1] < len(input[0]):
            paths.extend(dfs(input, new_pos, num+1, path))
    return paths

def trailheadRating(paths):
    trailheads = defaultdict(set)
    for path in paths:
        trailheads[path[0]].add(path[-1])
    return trailheads
    
def part1(input):
    return sum(len(t) for t in trailheadRating(find_paths(input)).values())

def part2(input):    
    return len(find_paths(input))

def main():
    input = open("input.txt", "r").read().splitlines()
    print(f"Part 1: {part1(input)}")
    print(f"Part 2: {part2(input)}")

if __name__ == "__main__":
    main()

File: Go/day10/AOC.go
package main

import (
	"fmt"
	"os"
	"strings"
	"utils"
)

type point struct {
	x, y int
}

func (p point) add(p2 point) point {
	return point{p.x + p2.x, p.y + p2.y}
}

func trailheadScore(hikingTrail map[point]rune, trailhead point) int {
	summits := 0
	seen := make(map[point]bool)
	directions := []point{{0, 1}, {0, -1}, {1, 0}, {-1, 0}}
	queue := []point{trailhead}

	for len(queue) > 0 {
		current := queue[0]
		queue = queue[1:]
		for _, dir := range directions {
			next := current.add(dir)
			if _, ok := hikingTrail[next]; !ok {
				continue
			}
			if hikingTrail[next] != hikingTrail[current]+1 {
				continue
			}

			if _, ok := seen[next]; ok {
				continue
			}

			seen[next] = true
			if hikingTrail[next] == '9' {
				summits++
			} else {
				queue = append(queue, next)
			}
		}
	}
	return summits
}

func part1(input []string) int {
	hikingTrail := make(map[point]rune)
	trailheads := []point{}
	for r, row := range input {
		for c, col := range row {
			hikingTrail[point{r, c}] = col
			if col == '0' {
				trailheads = append(trailheads, point{r, c})
			}
		}
	}
	scores := []int{}
	for _, trailhead := range trailheads {
		scores = append(scores, trailheadScore(hikingTrail, trailhead))
	}

	return utils.Sum(scores)
}

func trailheadRating(hikingTrail map[point]rune, trailhead point) int {
	trails := 0
	seen := map[point]int{trailhead: 1}
	directions := []point{{0, 1}, {0, -1}, {1, 0}, {-1, 0}}
	queue := []point{trailhead}
	for len(queue) > 0 {
		current := queue[0]
		queue = queue[1:]

		if hikingTrail[current] == '9' {
			trails += seen[current]
		}
		for _, dir := range directions {
			next := current.add(dir)
			if _, ok := hikingTrail[next]; !ok {
				continue
			}
			if hikingTrail[next] != hikingTrail[current]+1 {
				continue
			}
			if _, ok := seen[next]; ok {
				seen[next] += seen[current]
				continue
			}
			seen[next] = seen[current]
			queue = append(queue, next)
		}
	}
	return trails
}

func part2(input []string) int {
	hikingTrail := make(map[point]rune)
	trailheads := []point{}
	for r, row := range input {
		for c, col := range row {
			hikingTrail[point{r, c}] = col
			if col == '0' {
				trailheads = append(trailheads, point{r, c})
			}
		}
	}
	scores := []int{}
	for _, trailhead := range trailheads {
		scores = append(scores, trailheadRating(hikingTrail, trailhead))
	}

	return utils.Sum(scores)
}

func main() {
	input, err := os.ReadFile("input.txt")
	if err != nil {
		fmt.Print("Error reading file\n")
		panic(err)
	}

	lines := strings.Split(string(input), "\n")
	part1_ans := part1(lines)
	fmt.Printf("Part 1: %d\n", part1_ans)

	part2_ans := part2(lines)
	fmt.Printf("Part 2: %d\n", part2_ans)
}
